lib/puppet/http/client.rb

```diff
@@ -0,0 +1,140 @@
+      'User-Agent' => Puppet[:http_user_agent],
+    }.freeze
+    @default_ssl_context = ssl_context
+    @redirector = Puppet::HTTP::Redirector.new(redirect_limit)
+    @retry_after_handler = Puppet::HTTP::RetryAfterHandler.new(retry_limit, Puppet[:runinterval])
+  end
+
+  def connect(uri, ssl_context: nil, &block)
+    ctx = ssl_context ? ssl_context : default_ssl_context
+    site = Puppet::Network::HTTP::Site.from_uri(uri)
+    verifier = Puppet::SSL::Verifier.new(site.host, ctx)
+
+    @pool.with_connection(site, verifier) do |http|
+      if block_given?
+        handle_post_connect(uri, http, &block)
+      end
+    end
+  rescue Puppet::HTTP::HTTPError
+    raise
+  rescue => e
+    raise Puppet::HTTP::ConnectionError.new(_("Failed to connect to %{uri}: %{message}") % {uri: uri, message: e.message}, e)
+  end
+
+  def get(url, headers: {}, params: {}, ssl_context: nil, user: nil, password: nil, &block)
+    query = encode_params(params)
+    unless query.empty?
+      url = url.dup
+      url.query = query
+    end
+
+    request = Net::HTTP::Get.new(url, @default_headers.merge(headers))
+    apply_auth(request, user, password)
+
+    execute_streaming(request, ssl_context: ssl_context) do |response|
+      if block_given?
+        yield response
+      else
+        response.read_body
+      end
+    end
+  end
+
+  def put(url, headers: {}, params: {}, content_type:, body:, ssl_context: nil, user: nil, password: nil)
+    query = encode_params(params)
+    unless query.empty?
+      url = url.dup
+      url.query = query
+    end
+
+    request = Net::HTTP::Put.new(url, @default_headers.merge(headers))
+    request.body = body
+    request['Content-Length'] = body.bytesize
+    request['Content-Type'] = content_type
+    apply_auth(request, user, password)
+
+    execute_streaming(request, ssl_context: ssl_context) do |response|
+      response.read_body
+    end
+  end
+
+  def close
+    @pool.close
+  end
+
+  private
+
+  def execute_streaming(request, ssl_context:, &block)
+    redirects = 0
+    retries = 0
+
+    loop do
+      connect(request.uri, ssl_context: ssl_context) do |http|
+        http.request(request) do |nethttp|
+          response = Puppet::HTTP::Response.new(nethttp)
+          begin
+            Puppet.info("HTTP #{request.method.upcase} returned #{response.code} #{response.reason}")
+
+            if @redirector.redirect?(request, response)
+              request = @redirector.redirect_to(request, response, redirects)
+              redirects += 1
+              next
+            elsif @retry_after_handler.retry_after?(request, response)
+              interval = @retry_after_handler.retry_after_interval(request, response, retries)
+              retries += 1
+              if interval
+                Puppet.warning(_("Sleeping for %{interval} seconds before retrying the request") % { interval: interval })
+                ::Kernel.sleep(interval)
+                next
+              end
```

Comment:
If `Retry-After` is missing then `interval` is missing. We could either raise a ProtocolError or return the response as the final answer, which will not be "success?" Check how the current agent behaves.

Reply:
eg `raise Puppet::HTTP::ProtocolError, _("Location response header is missing") unless location`

Reply:
This matches the existing implementation. If the [`Retry-After` header is missing, we return the response](https://github.com/puppetlabs/puppet/blob/c501bd5042b84fb04d8e4119b5f78ad340290eed/lib/puppet/network/http/connection.rb#L243-L244), which causes us to [exit the retry loop](https://github.com/puppetlabs/puppet/blob/c501bd5042b84fb04d8e4119b5f78ad340290eed/lib/puppet/network/http/connection.rb#L196), and consider that to be our "final answer"